### PR TITLE
fix(driver): fall back to another EOA when a submission account is unusable

### LINKED ISSUE #4541: bug: EIP-7702 settlement submission lacks fallback mechanism in case of missing funds
### Problem
The driver can now use multiple EOAs for settlement submission with EIP-7702, but it does not check whether the selected EOA is actually usable before relying on it.

Auxiliary EOAs are selected from a FIFO queue. This means the selection order is fair, but it is not health-aware. If the next EOA in the queue has no funds, a bad nonce, or cannot replace a pending transaction, the driver can still select it.

Once that EOA is selected, the driver submits the settlement using that same signer. If submission fails because the selected EOA is unusable, the settlement fails. The driver does not try another available EOA.

For example:
- `EOA1` is next in the auxiliary queue
- `EOA1` has no native token for gas
- `EOA2` is funded and available
- the driver selects `EOA1`
- submission fails
- the driver does not retry with `EOA2`

So the issue is not that auxiliary EOAs are selected unfairly. The issue is that the selection is not health-aware, and there is no fallback when the selected EOA cannot submit.

### Impact
Impact: high.

A valid settlement can fail even when another usable EOA is available.

This reduces the reliability of EIP-7702 parallel settlement submission and can cause missed settlement opportunities. It also makes operations harder because an underfunded or nonce-stuck EOA can keep causing failures until operators notice and fix it manually.

### To reproduce (not tested manually)
1. Use latest `origin/main` at commit `754bb6a0a`.
2. Configure a solver with multiple EIP-7702 `submission_accounts`.
3. Make the next auxiliary EOA in the FIFO queue unusable, for example by leaving it without enough native token for gas.
4. Keep another auxiliary EOA funded and available.
5. Trigger a delegated settlement submission.
6. Observe that the driver selects the unusable EOA.
7. Observe that submission fails.
8. Observe that the driver does not retry the same settlement with the funded auxiliary EOA.

### Expected behaviour

Before selecting or using an EOA, the driver should check simple health signals where possible:
- native token balance is enough to pay for gas
- nonce state is okay (no pending transaction, e.g. rebalancing ERC20 assets while driver attempts to use that EOA for submission)
- account is not already known to be stuck or failing (some sort of blacklist?)

If the selected EOA fails for an account-specific reason, the driver should skip it and try another eligible EOA while the settlement is still valid. Account-specific failures include:
- insufficient funds for gas
- bad or stuck nonce
- replacement transaction underpriced etc.

The driver should not retry with another EOA for settlement-specific failures, such as:
- bad calldata
- target revert
- invalid on-chain state
- expired submission deadline

The driver should also temporarily mark unhealthy EOAs as unavailable so they are not selected again immediately.

### Screenshots/logs
Current behavior from code inspection on latest `origin/main` at commit `754bb6a0a`:

- Auxiliary EOAs are inserted into an `mpsc` channel in config order:
  https://github.com/cowprotocol/services/blob/754bb6a0a/crates/driver/src/domain/competition/mod.rs#L100-L112

- The driver always tries the direct solver EOA first. Delegated EOAs are only considered if the direct slot is busy:
  https://github.com/cowprotocol/services/blob/754bb6a0a/crates/driver/src/domain/competition/mod.rs#L124-L147

- The next delegated EOA is selected with `channel.recv()`. There is no balance or nonce health check before selecting it:
  https://github.com/cowprotocol/services/blob/754bb6a0a/crates/driver/src/domain/competition/mod.rs#L149-L160

- The selected submitter is used once for the settlement. If execution fails, the result is returned as `SubmissionError`; there is no retry with another EOA:
  https://github.com/cowprotocol/services/blob/754bb6a0a/crates/driver/src/domain/competition/mod.rs#L898-L921

- `Mempools::execute()` races configured mempools, but all attempts use the same selected `SubmissionMode`:
  https://github.com/cowprotocol/services/blob/754bb6a0a/crates/driver/src/domain/mempools.rs#L64-L96

- The selected signer is derived once from the prepared transaction and used for nonce lookup and submission:
  https://github.com/cowprotocol/services/blob/754bb6a0a/crates/driver/src/domain/mempools.rs#L133-L170
  https://github.com/cowprotocol/services/blob/754bb6a0a/crates/driver/src/domain/mempools.rs#L212-L220

- `send_transaction` failures are logged and returned as a generic submission error path:
  https://github.com/cowprotocol/services/blob/754bb6a0a/crates/driver/src/infra/mempool/mod.rs#L145-L179
  https://github.com/cowprotocol/services/blob/754bb6a0a/crates/driver/src/domain/mempools.rs#L633-L636

- Delegated EOAs are returned to the channel after the guard is dropped, so an unhealthy EOA can become selectable again later: https://github.com/cowprotocol/services/blob/754bb6a0a/crates/driver/src/domain/competition/mod.rs#L213-L222

### services version/commit hash and environment
Version / commit hash: `754bb6a0a`

### Additional context
The production runbook expects this kind of fallback behavior.

For underfunded or nonce-stuck auxiliary EOAs, the runbook expects:
- skip the bad auxiliary EOA
- use another funded auxiliary EOA if available
- fall back to direct submission when safe
- stop assigning new submissions to stuck nonce lanes until fixed

The driver should be updated so the implementation matches that operational model.

### PR DESCRIPTION
**Pull request by frdrckj**
# Description

EIP-7702 parallel submission selects a submission account from a FIFO pool. When the selected account cannot broadcast the transaction for an account-specific reason (no balance for gas, a stale nonce, or a pending tx that cannot be replaced), the settlement currently fails even though another funded account may be available — and the stuck account keeps getting picked by later settlements. This is the gap described in #4541.

This PR makes EIP-7702 submission health-aware: it retries a failed settlement from another account **and** temporarily benches the unhealthy account so subsequent settlements stop assigning to a stuck lane until it recovers.

# Changes

**Reactive fallback + classification**
* Add `mempools::Error::SubmitterUnusable(AccountFailure)` and a `classify_submission_failure` helper that recognises pre-broadcast node rejections (`insufficient funds`, `nonce too low`, `... underpriced`) across Geth- and Nethermind-style wording. `already known` and `nonce too high` are intentionally excluded.
* `Mempool::submit` returns this variant (matching the full `{:#}` error chain) for broadcast-time rejections instead of the generic `Other`.
* `Competition::process_settle_request` retries the settlement from another account on `SubmitterUnusable`. Spent accounts are held until the request finishes so each retry uses a different one, and a non-blocking `SubmitterPool::try_acquire` avoids deadlocking against concurrent settlements.

**Temporary quarantine — the "stop assigning to stuck lanes" half of #4541**
* On an account-specific failure the account is benched for `UNHEALTHY_ACCOUNT_COOLDOWN` (30s): it is withheld from the FIFO pool before being returned, so concurrent and subsequent settlements skip it until it recovers. It is retried automatically after the cooldown; a still-broken account is simply re-benched after one attempt. The direct solver EOA is never benched.

**Reliable propagation in multi-mempool setups**
* `Mempools::execute` races mempools with `select_ok`, which returns the *last* error when all fail — so a generic error (e.g. a timeout) from one mempool could mask a `SubmitterUnusable` from another and silently defeat the retry/quarantine. The race now surfaces `SubmitterUnusable` whenever any mempool reported one, unless a settlement-specific terminal failure (revert/expired) is present, which stays authoritative.

* Metrics and solver notifications account for the new outcome.

Retrying is safe against double-submission: a settlement that actually lands always returns `Ok`, so only failures (where nothing was broadcast) are ever retried. That is also why `already known` is excluded.

### Scope

This covers the two behaviours #4541 asks for: skip-and-retry on account-specific failures, and temporarily marking unhealthy EOAs unavailable so stuck lanes stop receiving new submissions. Proactive balance/nonce RPC pre-checks *before* selecting an account are intentionally **not** added — they put RPC round-trips on the latency-sensitive settlement path and are inherently racy (balance/nonce can change between check and submit), while the node's actual rejection (which this PR classifies and benches on) is the source of truth. Quarantine already satisfies the "account is not already known to be stuck" signal by physically removing benched accounts from selection. Happy to add pre-checks in a follow-up if preferred.

## How to test

```
RUST_MIN_STACK=3145728 cargo nextest run -p driver
```

Unit tests:
* `classifies_account_specific_submission_failures`, `classifies_non_geth_client_wording`, `does_not_classify_non_account_failures_as_account_specific` — the classifier in both directions (guards against retrying reverts, `already known`, rate limits, transport errors, `nonce too high`).
* `account_failure_is_surfaced_over_a_masking_race_error` — `SubmitterUnusable` wins over a masking generic error but defers to settlement-specific failures.
* `quarantined_account_is_withheld_from_the_pool`, `direct_slot_is_never_quarantined` — benching withholds an unhealthy account while leaving healthy ones available, and never benches the solver EOA.

Fixes #4541.

### CODE REVIEW

**Comment by frdrckj** [comment]
I have read the CLA Document and I hereby sign the CLA

File: crates/driver/src/domain/mempools.rs [hunk 1]
```diff
@@ -632,10 +651,88 @@ pub enum Error {
...
+    let message = message.to_lowercase();
+    // Match on keyword pairs rather than exact phrases so non-Geth clients are
+    // also covered (e.g. Nethermind's `SenderInsufficientFunds` / `NonceTooLow`,
+    // which omit the spaces Geth uses).
+    if message.contains("insufficient") && message.contains("funds") {
+        Some(AccountFailure::InsufficientFunds)
+    } else if message.contains("nonce") && message.contains("low") {
+        Some(AccountFailure::Nonce)
+    } else if message.contains("underpriced") {
+        Some(AccountFailure::Underpriced)
```
Comment (igorroncevic) [inline crates/driver/src/domain/mempools.rs]: Plain `transaction underpriced` is not always account-specific. It can mean the fee is below the node or network minimum, so retrying from another EOA will likely use the same fee path and can bench all submitters without fixing the submission.

Can we only classify replacement-underpriced errors here, such as `replacement transaction underpriced` or client equivalents, where the error shows this account has a stuck nonce?

File: crates/driver/src/domain/mempools.rs [hunk 2]
```diff
@@ -632,10 +651,88 @@ pub enum Error {
...
+    // A settlement-specific terminal failure is authoritative and must not be
+    // retried from another account, so it always wins over an account failure.
+    if matches!(
+        last_error,
+        Error::Revert { .. } | Error::SimulationRevert { .. } | Error::Expired { .. }
+    ) {
+        return last_error;
+    }
+    match account_failure {
+        Some(reason) => Error::SubmitterUnusable(reason),
```
Comment (igorroncevic) [inline crates/driver/src/domain/mempools.rs]: This can turn a post-broadcast error into a retryable account error. For example, one mempool can reject before broadcast, while another mempool already accepted the tx and later returns `Error::Other`, such as if the block stream ends after `send_transaction`.

In that case retrying from another EOA can double-submit the settlement while the first tx is still pending or already mined. Please keep `last_error` for errors that may happen after broadcast, and only override it with `SubmitterUnusable` when we know the failure happened before broadcast.

Comment (frdrckj) [reply crates/driver/src/domain/mempools.rs]: Oh yes, you're right, this is reachable. I'm sorry.

With multiple mempools, `execute()` races the lanes under `select_ok`, which only returns `Err` once all of them fail and then hands back whichever error finished last. So one lane can reject pre-broadcast with `SubmitterUnusable` while another already got past `send_transaction` and only later fails its wait loop with `Error::Other` (the block stream ending after submit, like you described). Since `race_error` only protects `Revert`/`SimulationRevert`/`Expired`, that post-broadcast `Other` gets overridden into `SubmitterUnusable`, and the retry loop resubmits from another EOA while the first tx is still in flight.

Fix: set a shared flag the moment any lane's `send_transaction` returns `Ok`, and only let `execute()` surface `SubmitterUnusable` when that flag is false, i.e. when nothing was put on the wire. If any lane broadcast, we keep the real error and never signal a retry. The pre-broadcast case still works as before (flag stays false, we surface `SubmitterUnusable` and retry).

I'll gate on the broadcast flag rather than on `last_error`'s type. `select_ok` can return the pre-broadcast `SubmitterUnusable` as `last_error` itself if that lane finishes last, even though another lane already broadcast, so checking the error alone isn't enough. Keying off "did anything broadcast" covers that.

On tests: `race_error` picks up the new broadcast arg, so `account_failure_is_surfaced_over_a_masking_race_error` needs updating. The existing `connection reset` assertion stays valid as the pre-broadcast case (flag false, still surfaces `SubmitterUnusable`). I'll add a post-broadcast case (flag true) asserting the `Other` is preserved, plus one for the corner where `last_error` is itself a `SubmitterUnusable` while another lane broadcast.

File: crates/driver/src/domain/mempools.rs (hunk 1 above)
Comment (frdrckj) [reply crates/driver/src/domain/mempools.rs]: Right now it treats anything containing `underpriced` as account-specific, so a plain `transaction underpriced` from the node's fee floor gets the same treatment as `replacement transaction underpriced`. In the global case the retry just recomputes the same fee on the next EOA and fails again, benching each one for the cooldown without fixing anything. Exactly what you described.

Fix: I narrowed it to require both `replacement` and `underpriced`, so the global forms fall through to `None`. `replacement transaction underpriced` still classifies, and since it's substring matching on the lowercased message, spaceless client variants carrying both tokens are still covered. Only the plain/global strings drop out.

I also added a negative test asserting `transaction underpriced` and `max fee per gas less than block base fee` return `None`, since nothing guarded that before.

File: crates/driver/src/domain/mempools.rs (hunk 2 above)
Comment (igorroncevic) [reply crates/driver/src/domain/mempools.rs]: Thanks, the `any_broadcast` flag fixes the post-broadcast case from this thread.

I think we still need one more guard here: if another lane returns an ambiguous non-account error before confirmed broadcast, like timeout, connection reset, `already known`, or `nonce too high`, that should stay non-retryable. Please track an non-retryable failure from any lane and only return `SubmitterUnusable` when no lane broadcasted and no such error was seen.

File: crates/driver/src/domain/mempools.rs [hunk 3]
```diff
@@ -745,4 +877,123 @@ mod tests {
...
+        );
+        assert_eq!(classify_submission_failure("NonceTooLow"), Some(Nonce));
+        assert_eq!(
+            classify_submission_failure("ReplacementTransactionUnderpriced"),
+            Some(Underpriced)
+        );
+    }
+
+    #[test]
+    fn account_failure_is_surfaced_over_a_masking_race_error() {
```
Comment (igorroncevic) [inline crates/driver/src/domain/mempools.rs:1031]: Fallback is only safe when we know no tx was sent.

If we get an unclear error like timeout, connection reset, `already known`, or `nonce too high`, we cannot be sure no tx is pending.

In that case, please keep the error non-retryable. Only return `SubmitterUnusable` when all active mempools failed before sending.

File: crates/driver/src/domain/mempools.rs [hunk 4]
```diff
@@ -218,6 +244,9 @@ impl Mempools {
                 nonce,
             )
             .await?;
+        // The tx is now on the wire. Record it so a sibling mempool's pre-broadcast
+        // account failure can't trigger a retry that double-submits this settlement.
+        broadcasted.store(true, std::sync::atomic::Ordering::SeqCst);
```
Comment (igorroncevic) [inline crates/driver/src/domain/mempools.rs:279]: Please check the signer balance before this point.

If the balance is too low, return `SubmitterUnusable(InsufficientFunds)` so another account can be tried.

File: crates/driver/src/domain/competition/mod.rs [hunk 5]
```diff
@@ -912,17 +971,51 @@ impl Competition {
...
-        let executed = self
-            .mempools
-            .execute(&settlement, submission_deadline, &mode)
-            .await;
+        let mut spent = Vec::new();
+        let executed = loop {
+            let mode = guard.submission_mode();
+            let executed = self
+                .mempools
+                .execute(&settlement, submission_deadline, &mode)
```
Comment (igorroncevic) [inline crates/driver/src/domain/competition/mod.rs]: Please check the deadline right before each submit attempt. `acquire()` can wait, so the deadline may pass before this line runs.

File: crates/driver/src/domain/competition/mod.rs [hunk 6]
```diff
@@ -209,13 +249,32 @@ impl SubmitterGuard {
             GuardInner::Dropped => unreachable!(),
         }
     }
+
+    /// Bench the underlying delegated account: once this guard drops, the
+    /// account is withheld from the pool for `cooldown` so concurrent and
+    /// subsequent settlements stop selecting a stuck or underfunded lane until
+    /// it recovers. No-op for the direct solver EOA, which is never benched.
+    fn quarantine(&mut self, cooldown: Duration) {
```
Comment (igorroncevic) [inline crates/driver/src/domain/competition/mod.rs]: Please pass the failure reason into quarantine. The logs should show which account was benched, why it was benched, and for how long.

File: crates/driver/src/domain/competition/mod.rs [hunk 7]
```diff
@@ -912,17 +971,51 @@ impl Competition {
...
+                break executed;
+            }
+            let Some(next) = self.submitter_pool.try_acquire() else {
+                // No other account currently free; give up and report the error.
+                break executed;
+            };
+            tracing::warn!(
+                ?mode,
+                "submission account unusable, retrying from another account"
+            );
```
Comment (igorroncevic) [inline crates/driver/src/domain/competition/mod.rs]: Please make this retry log more useful. It should show the failed account, the next account, and the reason for retrying.

File: crates/driver/src/domain/competition/mod.rs [hunk 8]
```diff
@@ -1102,3 +1195,84 @@ pub enum Error {
     #[error("could not parse the request")]
     MalformedRequest,
 }
+
+#[cfg(test)]
+mod submitter_pool_tests {
```
Comment (igorroncevic) [inline crates/driver/src/domain/competition/mod.rs:1231]: These pool tests are useful, but we also need tests for the full retry flow.

Please cover direct-to-delegated fallback, delegated-to-direct fallback, no account available, deadline reached, unsafe errors, and tx already sent.

File: crates/driver/src/domain/mempools.rs [hunk 9]
```diff
@@ -632,10 +665,109 @@ pub enum Error {
...
     Other(#[from] anyhow::Error),
 }
 
+/// Account-specific reasons a node rejects a transaction at submission time,
+/// before it is broadcast.
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+pub enum AccountFailure {
+    InsufficientFunds,
+    Nonce,
+    Underpriced,
```
Comment (igorroncevic) [inline crates/driver/src/domain/mempools.rs]: Can we rename this to `ReplacementUnderpriced`? Plain `transaction underpriced` should not trigger fallback. Only replacement underpriced should.

File: crates/driver/src/domain/competition/mod.rs
Comment (igorroncevic) [inline crates/driver/src/domain/competition/mod.rs:1056]: Please keep solver notifications outside the retry loop.

Each retry attempt should be logged inside the loop, but `notify::executed(...)` should only run once, after the loop finishes with the final result.

  So the split should be:
  - inside the loop: log each failed account and retry
  - after the loop: send one final solver notification

  Please add a small test or check for this if it is easy.

**Review by igorroncevic — COMMENTED** [review]
A few minor things

File: crates/driver/src/domain/mempools.rs (hunk 2 above)
Comment (frdrckj) [reply crates/driver/src/domain/mempools.rs]: I'll add a second shared flag in `execute()`, `saw_nonretryable`, set for any lane result that isn't `SubmitterUnusable` and isn't `Disabled` (Disabled is a clean skip that never touched the network). `race_error` then returns `SubmitterUnusable` only when nothing broadcast and that flag stays false, so every active mempool failed cleanly before sending. Otherwise it keeps the real error, and strips a `SubmitterUnusable` that `select_ok` happened to return last so it can't leak a retry.

Tests: the existing `connection reset` case in `account_failure_is_surfaced_over_a_masking_race_error` now flips so it no longer triggers a retry, since that's exactly the ambiguous sibling you describe. I'll add a case for the corner where the last error is itself `SubmitterUnusable` while the flag is set, and keep the case where every lane failed cleanly and we still retry.

File: crates/driver/src/domain/competition/mod.rs (hunk 5 above)
Comment (frdrckj) [reply crates/driver/src/domain/competition/mod.rs]: I'll move the deadline check to the top of the loop so it runs before each execute(), including the first one. A request whose window passed while it waited for a slot fails fast without submitting, and the current post-failure check folds into the same top-of-loop check.

File: crates/driver/src/domain/competition/mod.rs (hunk 6 above)
Comment (frdrckj) [reply crates/driver/src/domain/competition/mod.rs]: Will do. I'll pass the failure reason into quarantine() and store it on the guard, then include it in the bench log so it shows which account was benched, why (insufficient_funds, nonce, or replacement_underpriced), and for how long.

File: crates/driver/src/domain/competition/mod.rs (hunk 7 above)
Comment (frdrckj) [reply crates/driver/src/domain/competition/mod.rs]: I'll expand the retry log to include the failed account, the next account it falls back to, and the reason.

File: crates/driver/src/domain/competition/mod.rs
Comment (frdrckj) [reply crates/driver/src/domain/competition/mod.rs:1056]: notify::executed already runs once, after the loop, with the final result; only the per-attempt retries log inside the loop, so the structure already matches what you describe. When I extract the retry loop into a helper, notify::executed stays in process_settle_request and is called once on the helper's returned result, so there is no path that double-notifies or skips it. Asserting exactly one notification at the unit level would need a mock notify sink driving process_settle_request (a full Competition harness), which is not set up today. The call site is single by construction; happy to add that harness test if you want it.

File: crates/driver/src/domain/competition/mod.rs (hunk 8 above)
Comment (frdrckj) [reply crates/driver/src/domain/competition/mod.rs:1231]: I'll extract the retry loop into a small helper, submit_with_fallback(pool, guard, cooldown, deadline_reached, submit), that takes the submit step as a closure, so it can be unit-tested without the network. Tests will cover direct to delegated fallback, delegated to direct fallback, no account available, deadline reached (fails fast before submitting), an unsafe or ambiguous error (no retry), and success on the first attempt (no retry). These assert the helper's own contract; notify::executed stays in process_settle_request and is covered there

File: crates/driver/src/domain/mempools.rs (hunk 9 above)
Comment (frdrckj) [reply crates/driver/src/domain/mempools.rs]: I'll rename AccountFailure::Underpriced to ReplacementUnderpriced

File: crates/driver/src/domain/mempools.rs (hunk 3 above)
Comment (frdrckj) [reply crates/driver/src/domain/mempools.rs:1031]: Covered by the saw_nonretryable guard.
race_error now returns SubmitterUnusable only when no lane broadcast and no lane saw an ambiguous error (timeout, connection reset, already known, nonce too high), so any of those from any lane keeps the result non-retryable.

File: crates/driver/src/domain/mempools.rs (hunk 4 above)
Comment (frdrckj) [reply crates/driver/src/domain/mempools.rs:279]: I now fetch the signer's balance and compare it to required_balance (the gas limit times the submission max_fee_per_gas, reusing the existing settlement helper). If the account cannot cover the gas, submit() returns SubmitterUnusable(InsufficientFunds) without sending, so the fallback picks another account and benches this one. A failed balance lookup is non-authoritative: I log it and proceed, letting the node decide rather than blocking submission on an RPC hiccup. This adds one eth_getBalance per submission attempt, and the reactive path still catches the case where the balance drops between the check and the send.

**Comment by igorroncevic** [comment]
Fixes look good so far. I'm going to have to ask @cowprotocol/backend squad to step in for a more detailed look. Thanks for your contribution!

**Comment by frdrckj** [comment]
Thanks Igor, appreciate the review! Happy to work through any feedback from the backend squad.

File: crates/driver/src/domain/competition/mod.rs [hunk 10]
```diff
@@ -64,6 +64,17 @@ type Balances = HashMap<BalanceGroup, order::SellAmount>;
...
 
+/// How long an EIP-7702 submission account is benched after it fails to
+/// broadcast for an account-specific reason (no gas for the tx, a stuck nonce,
+/// a pending tx that can't be replaced). While benched, the account is held out
+/// of the selection pool, so the driver stops assigning settlements to a stuck
+/// lane until it recovers. The account is retried automatically once the
+/// cooldown elapses; a still-broken account is simply re-benched after a single
+/// failed attempt.
+// ponytail: a fixed cooldown is enough; lift to driver config only if operators
+// need per-network tuning.
```
Comment (squadgazzz) [inline crates/driver/src/domain/competition/mod.rs]: This looks too verbose + some AI slop.
```suggestion
/// How long a submission account sits out of the pool after an
/// account-specific failure (no gas, stuck nonce). It rejoins once the
/// cooldown elapses and is re-benched if it fails again.
```

File: crates/driver/src/domain/mempools.rs [hunk 11]
```diff
@@ -209,6 +253,32 @@ impl Mempools {
...
+                );
+                return Err(Error::SubmitterUnusable(AccountFailure::InsufficientFunds));
+            }
+            Ok(_) => {}
+            Err(err) => tracing::warn!(
+                ?signer,
+                ?err,
+                "could not check submission account balance before submitting"
+            ),
+        }
```
Comment (squadgazzz) [inline crates/driver/src/domain/mempools.rs]: IIUC the description says proactive balance pre-checks were left out on purpose, but this block adds exactly that check, one `eth_getBalance` per mempool before every broadcast, happy path included. The classifier already turns the node's own insufficient funds rejection into the same SubmitterUnusable, so the pre-check spends an RPC on every submission to save one doomed broadcast in the rare failure case. Probably makes sense to drop this block and let the node's rejection be the signal, which would also match the description, wdyt?

File: crates/driver/src/domain/mempools.rs [hunk 12]
```diff
@@ -632,10 +709,126 @@ pub enum Error {
...
+/// double-submit; or `saw_nonretryable` is set, meaning a mempool failed before
+/// broadcast with an ambiguous error (timeout, connection reset, `already
+/// known`, `nonce too high`) whose tx might still be live, which is likewise
+/// unsafe to retry.
+fn race_error(
+    last_error: Error,
+    account_failure: Option<AccountFailure>,
+    broadcasted: bool,
+    saw_nonretryable: bool,
+) -> Error {
```
Comment (squadgazzz) [inline crates/driver/src/domain/mempools.rs:780]: Did I get it correctly, that a terminal failure only wins the race when it happens to be the error `select_ok` returned last? If one lane reports SimulationRevert and another lane's SubmitterUnusable finishes last, the retry is still correctly suppressed (the revert sets saw_nonretryable), but the surfaced error becomes `Other(...)`, so the solver is notified Fail instead of SimulationRevert and the metrics label shifts the same way. Did you consider tracking a terminal error from any lane, or is the label drift acceptable here?

File: crates/driver/src/domain/mempools.rs (hunk 11 above)
Comment (frdrckj) [reply crates/driver/src/domain/mempools.rs]: yes, the code and the description is different. Igor asked for this pre-check earlier, which is why it's in. Functionally the reactive path already covers the case: the classifier turns the node's insufficient-funds rejection into the same `SubmitterUnusable`, so an underfunded account still gets skipped and benched without the pre-check. The only thing the pre-check buys is skipping the doomed broadcast, plus a bit of robustness if a sibling relay returns something ambiguous instead of a clean rejection. @igorroncevic, what do you think? should we drop it?

File: crates/driver/src/domain/mempools.rs (hunk 12 above)
Comment (frdrckj) [reply crates/driver/src/domain/mempools.rs:780]: oh yes, i havent consider it yet. I can track a terminal error from any lane, the same way I capture the account failure, and return it ahead of the account-failure path so `Revert`/`SimulationRevert`/`Expired` always win regardless of race order. Adding a test for the two-lane case (one reverts, the other reports an account failure last).

File: crates/driver/src/domain/competition/mod.rs (hunk 8 above)
Comment (igorroncevic) [reply crates/driver/src/domain/competition/mod.rs:1231]: I’ll leave this coverage question for the backend team to decide.

File: crates/driver/src/domain/mempools.rs (hunk 11 above)
Comment (igorroncevic) [reply crates/driver/src/domain/mempools.rs]: I will respect whatever backend squad recommends. @frdrckj

Comment (frdrckj) [reply crates/driver/src/domain/mempools.rs]: I'll drop it then, the code will match the description again too

**Comment by igorroncevic** [comment]
cc @cowprotocol/backend
